doc: build: updates to build from source

ranukund opened the pull request:
# Description

Addressing the recommendations in the [DOC 13142](https://jira.devtools.intel.com/browse/DOC-13142)

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   ~~~
+
+3. Build the library
+   ~~~sh
```
Provide the same description for this step as we have in the previous section (line 91). Provide the command for macOS as well

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   cmake .. -DDNNL_AARCH64_USE_ACL=ON
+   ~~~
+
+3. Build the library
```
Provide the same description for this step as we have in the previous section (line 91). Provide the command for macOS as well

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   cmake --build . --config=Release
+   ~~~
+
+   @note Currently, the oneDNN build system has limited support for multi-config
```
The note on the lines 204-213 should be clarified and probably rewritten. Maybe it should be separated into 2 different notes

ranukund replied:
We can have a bulleted list for enhanced clarity.

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   both `Debug` and `Release` builds), but it must refer to the same build type
+   (`Release`, `Debug`, etc.) as selected with the `CMAKE_BUILD_TYPE` option.
+
+   @note You can also open `oneDNN.sln` to build the project from the
```
Clarify that this is another option for step 3 Build the library. Maybe we should move this note higher and make it step Option B instead of keeping it in the note.   
```suggestion
   @note Alternatively, you can open `oneDNN.sln` to build the project from the
```

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   # Some generators, like Unix Makefiles, might default to single-threaded
+   # compilation. Parallelization can be controlled with:
+   cmake --build . --parallel $(nproc)
+   ~~~
```
Proceed with 'Verify the Build' (relative link)

ranukund replied:
As mentioned earlier, let's avoid repeating the same information at every stage.

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   ~~~sh
+   cmake --build . --parallel $(nproc)
+   ~~~
+
```
Proceed with 'Verify the Build' (relative link)

ranukund replied:
Let's not repeat the same information at every stage. Let's wait for the tabbed approach to be implemented and the guide can be enhanced further with subsequent PRs.

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   ~~~sh
+   cmake --build . --parallel $(nproc)
+   ~~~
+
```
Proceed with 'Verify the Build' (relative link)

ranukund replied:
Let's not repeat the same information at every stage. Let's wait for the tabbed approach to be implemented and the guide can be enhanced further with subsequent PRs.

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   ~~~sh
+   cmake --build . --parallel $(nproc)
+   ~~~
+
```
Proceed with 'Verify the Build' (relative link)

ranukund replied:
Let's not repeat the same information at every stage. Let's wait for the tabbed approach to be implemented and the guide can be enhanced further with subsequent PRs.

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+
+   @note You can also open `oneDNN.sln` to build the project from the
+   Microsoft Visual Studio IDE.
+
```
Proceed with 'Verify the Build' (relative link)

Marina-Zubova replied:
not sure how tabs can resolve this, do we know how the document will look like after applying tabs? I'd like to understand better what will change

ranukund replied:
Until tabs are implemented, we can add numbering for each step as mentioned above.

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   cd build
+
+   cmake .. -G Ninja ^
          -DDNNL_CPU_RUNTIME=SYCL ^
```
update the prefix
```suggestion
         -DONEDNN_CPU_RUNTIME=SYCL ^
```

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+
+   cmake .. -G Ninja ^
          -DDNNL_CPU_RUNTIME=SYCL ^
          -DDNNL_GPU_RUNTIME=SYCL
```
```suggestion
         -DONEDNN_GPU_RUNTIME=SYCL
```

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+2. Generate the build system
+   ~~~sh
+   mkdir -p build ; cd build
+   cmake .. -DDNNL_AARCH64_USE_ACL=ON
```
```suggestion
   cmake .. -DONEDNN_AARCH64_USE_ACL=ON
```

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   ~~~sh
+   mkdir -p build ; cd build
+   cmake .. -DDNNL_CPU_RUNTIME=SYCL \
          -DDNNL_GPU_RUNTIME=SYCL
```
```suggestion
         -DONEDNN_GPU_RUNTIME=SYCL
```

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+2. Generate the build system
+   ~~~sh
+   mkdir -p build ; cd build
+   cmake .. -DDNNL_CPU_RUNTIME=SYCL \
```
```suggestion
   cmake .. -DONEDNN_CPU_RUNTIME=SYCL \
```

Marina-Zubova reviewed (changes requested):
This is the first round of DX review, left comments and also suggestions where possible

ranukund commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   Full list of options can be found [here](https://cmake.org/cmake/help/latest/manual/cmake.1.html#build-a-project).
 
-#### GCC, Clang, or Intel oneAPI DPC++/C++ Compiler
+You can build the library on Linux, macOS, or Windows using the compiler of your choice.
```
This is added under Build the Library option. From the right hand side navigation menu, it should be understandable what the next step is. The tabbed approach can be implemented to enhance the usability which will reduce the page scroll time.

ranukund replied:
From my local build, the right hand side navigation menu looks as follows: 

![image](https://github.com/user-attachments/assets/681363b2-2417-4c2e-84dd-b712a71e9fd5)

ranukund commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
-# export CXX=clang++
+1. Set up the environment for the compiler
+   ~~~sh
+   # Uncomment the following lines to build with GCC
```
As mentioned earlier, we can enhance usability by implementing tabbed approach.

ranukund commented:
> This is the first round of DX review, left comments and also suggestions where possible

A lot of comments are around the commands. Adding @spalicki for review.

spalicki commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   ~~~
+   In most cases, both `-B` and `-S` options are skipped with the assumption that
+  `<path-to-build>` is the current folder and `<path-to-source>` is the one
+   higher in the tree:
```
Again, not usual.

spalicki commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   mkdir -p build ; cd build
+   cmake .. [<options>]
+   ~~~
+   The following are a few useful options defined by CMake:
```
This section defines CMake options not one defined by oneDNN, those are linked below. We should most likely then move the oneDNN options up front.

mzhukova replied:
I think including all the oneDNN options here in addition to CMake ones would crowd page way too much, we have a Build Options page for a reason (it is really big table!)

spalicki replied:
Not all just 3/4 most common ones CPU/GPU/TEST and then point to the whole table?

mzhukova replied:
CMake variables are not oneDNN build options (== oneDNN supported functionality), so I think we should leave it up to CMake documentation to list them. Here we just mention a few frequently used ones for user convenience.

We already provide a link to CMake documentation at the beginning of this section, but we could also provide link to CMake options (https://cmake.org/cmake/help/latest/manual/cmake.1.html#options) and CMake variables (https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html) if you think that is necessary..

Marina-Zubova replied:
Seems like we agreed on CMake arguments, but what about ONEDNN ones, are we going to provide the most common ones in this guide to support new users?

spalicki commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+
+3. Build the library
+   ~~~bat
+   cmake --build . --config=Release
```
If we want to be consistent, we can add:
```
cmake --build . --config=Release --parallel %NUMBER_OF_PROCESSORS%
```

spalicki commented:
> > This is the first round of DX review, left comments and also suggestions where possible
> 
> A lot of comments are around the commands. Adding @spalicki for review.

I have added my comments to existing review.

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   cmake -B <path-to-build> [-S <path-to-source>] [<options>]
+   ~~~
+   In most cases, both `-B` and `-S` options are skipped with the assumption that
+  `<path-to-build>` is the current folder and `<path-to-source>` is the one
```
it can be replaced with any usual path to oneDNN build and source that oneDNN engineers provide. Do you have any suggestions?

vpirogov replied:
Maybe for the sake of documentation we can assume that all actions happen in the folder where oneDNN was cloned?

vpirogov replied:
This looks like common practice:
* https://github.com/pytorch/pytorch?tab=readme-ov-file#get-the-pytorch-source
* https://www.tensorflow.org/install/source#download_the_tensorflow_source_code

Marina-Zubova replied:
If we don't have usual installation and build paths, let's not mention it then. We do have 'cd oneDNN' command included (at the very top of the document), so assuming developer is in this folder

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   ~~~sh
+   cd <path-to-onednn-source>
+   mkdir -p build ; cd build
+   cmake .. [<options>]
```
my comment above where I was explaining the issue has been removed for some reason. The idea was that it's confusing to see 2 cmake commands here while you only need to run it once

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
-configuration options defined by oneDNN.
+   - `CMAKE_PREFIX_PATH` to specify directories to be searched for the
+    dependencies located at non-standard locations.
 
```
To support developers, voting for listing of the most common ONEDNN cmake arguments with most common values. If more details are needed, developer can refer to Build option doc, but the most common arguments should be called out

mzhukova replied:
tbh doesn't make sense to me, the most common build configuration of the project is probably what defined by default values of all build options, so you actually don't need to specifically call them out, because this is what you get by not specifying anything, right..?

spalicki replied:
The default for `ONEDNN_GPU_RUNTIME` is `NONE`, so people might be confused why nothing is working on GPU when built with just `cmake ..`.

mzhukova replied:
closing this thread as a duplication

mzhukova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
+   The oneDNN build system is based on [CMake](https://cmake.org/cmake/help/latest/manual/cmake.1.html). 
+   Use the following command to generate a build system:
+   ~~~sh
+   cmake -B <path-to-build> [-S <path-to-source>] [<options>]
```
I agree with @spalicki, I think we should leave the simple / common one and add -B, -S to the list of useful options (especially since everywhere below we're using that one)

ranukund commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
+## Step 2. Build the Library
 
-### Set up the environment for the compiler
+### Prerequisites
```
I see that introducing Prerequisites can be problematic. I plan to remove the header and add just the sentence about ensuring the software dependencies are in place. I usually look at the rendered output before committing changes to a PR. However, this time around, the commit got accidentally pushed.

ranukund commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
-cmake .. -DCMAKE_SYSTEM_NAME=Linux \
+1. Set up the environment for the compiler
+
+Configure your operating system's 
```
fixed in my next commit

ranukund commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
 ~~~
 
-## Build the Library
+## Step 2. Build the Library
```
I see this is confusing. As mentioned in my earlier comment. I plan to remove the header and add just the sentence about ensuring the software dependencies are in place.

ranukund commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
+
+3. Build the library
+
+   Generators such as Unix Makefiles might default to single-threaded 
```
Fixed in my next commit

mzhukova replied:
I wonder if instead of having "Multi-threaded compilation is recommended for a faster build process. Use the --parallel option to specify the number of parallel jobs." in each sub-section for different compilers, we just have this note once in ~generic section for better readability? (but keep `--parallel` everywhere anyway)

Marina-Zubova replied:
looks good to me, assuming we still keep similar messages in further specific sections

ranukund commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
+   ~~~bat
+   cmake --build .
+   ~~~
+
```
fixed in the next commit, thanks

ranukund commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
-After building the library, you can run a predefined test set using:
 ~~~sh
 ctest
 ~~~
```
fixed, thanks

mzhukova commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
+   @note Alternatively, you can open `oneDNN.sln` to build the project from the
+   Microsoft Visual Studio IDE.
+
+#### Use Intel oneAPI DPC++/C++ Compiler with SYCL Runtime
```
double checked after latest commit, but warning still exists and section is still misplaced

ranukund replied:
I'm not sure what's causing the issue. I need to investigate. I saw the warning still exists.

mzhukova replied:
oh sure, you just re-requested review, so I assumed everything got addressed

mzhukova commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
+   cmake --build .
+   ~~~
+
+## Step 4. Validate the Build
```
>But wouldn't the user like to validate the build right after building the library?

User might (or not) want to, but they *don't have to* do validation in order to build or install a library, that was my point. 
And therefore it probably makes sense to also mark it as optional or separate it from the other steps (if we're doing this to documentation step).

> Build documentation is optional but is it tied to Installation of the library?

Installation is a separate step that is basically copying stuff to a dedicated folder, so if you have reference documentation built already, then on installation step it would be copied to <install-folder>/share/doc/..., otherwise (if documentation was not built previously) doc folder would only include README, not the complete reference guide AFAIK.

spalicki commented on doc/build/build.md:
```diff
@@ -1,271 +1,278 @@
 
 ~~~sh
 git clone https://github.com/uxlfoundation/oneDNN.git
+cd oneDNN
```
I would leave it and assume that every other command is executed inside `oneDNN`, then remove the `cd <path-to-source>` below.

spalicki commented:
@ranukund If we want to make it easy to parse for new developers, how about we streamline the documentation reorganize the structure and remove all the duplicated commands and notes? Something similar to #3395

Marina-Zubova commented on doc/build/build.md:
```diff
@@ -8,264 +9,256 @@ or clone [the repository](https://github.com/uxlfoundation/oneDNN.git).
 
-Ensure that all [software dependencies](https://github.com/uxlfoundation/oneDNN#requirements-for-building-from-source)
-are in place and have at least the minimal supported version.
+1. Set up the environment for the compiler
```
okay, let's address the prerequisites in a separate PR. It makes sense to still keep the step 1. Set up the environment for the compiler here for consistency. But instead of pointing to prerequisites (this will be moved in the future), we can inform about that on this step you may need to set environment variables specific to your compiler.

ranukund commented:
Should I handle this PR in phases, cherry-pick the latest stable version, and publish the approved changes? In the next PR, I can address the remaining comments if required. @mzhukova @spalicki - please let me know what suits.

mzhukova commented on doc/build/build.md:
```diff
@@ -1,276 +1,305 @@
+   ~~~sh
+   mkdir -p build ; cd build
+   cmake .. -DONEDNN_CPU_RUNTIME=SYCL \
+         -DONEDNN_GPU_RUNTIME=SYCL
```
(minor) fix alignment?

mzhukova commented on doc/build/build.md:
```diff
@@ -294,26 +323,30 @@ When using the `/opt/intel/oneapi/setvars.sh` script from the Intel oneAPI toolk
-~~~
+Install the library, headers, and documentation.
+
+@note
```
(minor) seems like everywhere else we have a note after the actual command, not before

ranukund replied:
The usability study participants missed the information about having administrator privileges and running using the sudo command when installing in the default directory. Therefore, had to move this content up before the command. I also think we need not use the note admonition.

mzhukova commented:
> Should I handle this PR in phases, cherry-pick the latest stable version, and publish the approved changes? In the next PR, I can address the remaining comments if required. @mzhukova @spalicki - please let me know what suits.

hi @ranukund, I think there are just a few threads + minor comments left, it would be great to have them resolved before merging so to not loose track and also to keep Build page update self-contained. Thanks!

ranukund commented:
> > Should I handle this PR in phases, cherry-pick the latest stable version, and publish the approved changes? In the next PR, I can address the remaining comments if required. @mzhukova @spalicki - please let me know what suits.
> 
> hi @ranukund, I think there are just a few threads + minor comments left, it would be great to have them resolved before merging so to not loose track and also to keep Build page update self-contained. Thanks!

Hi @mzhukova I've addressed your comments.

Marina-Zubova reviewed (changes requested):
Added a few more comments and minor suggestions. The document looks much better, we have a good progress!

ranukund commented on doc/build/build.md:
```diff
@@ -294,26 +321,29 @@ When using the `/opt/intel/oneapi/setvars.sh` script from the Intel oneAPI toolk
-cmake variable. When installing in the default directory, the above command
-needs to be run with administrative privileges using `sudo` on Linux/Mac or a
-command prompt run as administrator on Windows.
+
```
We can address this in the next PR.

ranukund commented on doc/build/build.md:
```diff
@@ -1,276 +1,303 @@
 
-- Set up the environment for the compiler
+   ~~~sh
+   cmake --build <path-to-build> --parallel<numberofprocessors> [<options>]
```
addressed, thanks

ranukund commented on doc/build/build.md:
```diff
@@ -294,26 +321,29 @@ When using the `/opt/intel/oneapi/setvars.sh` script from the Intel oneAPI toolk
 Make sure the correct oneDNN library is present in
```
Let's work on this in the next PR

mzhukova commented on doc/build/build.md:
```diff
@@ -289,31 +320,34 @@ Another useful option is `--output-on-failure`, which will print verbose output
-`LD_LIBRARY_PATH` is set to include the oneDNN library path from the installation.
-Make sure the correct oneDNN library is present in
-`LD_LIBRARY_PATH` by setting it explicitly if needed.
+When using the `/opt/intel/oneapi/setvars.sh` script from the Intel oneAPI toolkit,
```
IPLDT (linter) is currently unhappy about Toolkit -> toolkit change, if this one is indeed correct, I could update the hits.db after this PR is merged
@ranukund please let me know
